pallets/baby-liminal/src/lib.rs

```diff
@@ -114,40 +154,61 @@ pub mod pallet {
         #[pallet::call_index(0)]
         #[pallet::weight(T::WeightInfo::store_key(key.len() as u32))]
         pub fn store_key(
-            _origin: OriginFor<T>,
+            origin: OriginFor<T>,
             identifier: VerificationKeyIdentifier,
             key: Vec<u8>,
         ) -> DispatchResult {
-            Self::bare_store_key(identifier, key).map_err(|e| e.into())
+            Self::bare_store_key(origin, identifier, key).map_err(|e| e.into())
         }
 
         /// Deletes a key stored under `identifier` in `VerificationKeys` map.
         ///
-        /// Can only be called by a root account.
+        /// Returns the deposit locked. Can only be called by the key owner.
         #[pallet::call_index(1)]
-        #[pallet::weight(T::DbWeight::get().writes(1))]
+        #[pallet::weight(T::WeightInfo::delete_key(T::MaximumVerificationKeyLength::get()))]
         pub fn delete_key(
             origin: OriginFor<T>,
             identifier: VerificationKeyIdentifier,
         ) -> DispatchResult {
-            ensure_root(origin)?;
+            let who = ensure_signed(origin).map_err(|_| Error::<T>::BadOrigin)?;
+            let owner = VerificationKeyOwners::<T>::get(identifier)
+                .ok_or(Error::<T>::UnknownVerificationKeyIdentifier)?;
+
+            ensure!(who == owner, Error::<T>::NotOwner);
+
+            let deposit = VerificationKeyDeposits::<T>::take((&owner, &identifier)).unwrap(); // cannot fail since the key has owner and owner must have made a deposit
+            T::Currency::unreserve(&owner, deposit);
+
             VerificationKeys::<T>::remove(identifier);
-            Self::deposit_event(Event::VerificationKeyDeleted);
+            Self::deposit_event(Event::VerificationKeyDeleted(who, identifier));
             Ok(())
         }
 
         /// Overwrites a key stored under `identifier` in `VerificationKeys` map with a new value `key`
         ///
         /// Fails if `key.len()` is greater than `MaximumVerificationKeyLength`.
-        /// Can only be called by a root account.
+        /// Can only be called by the original owner of the key.
         #[pallet::call_index(2)]
         #[pallet::weight(T::WeightInfo::overwrite_key(key.len() as u32))]
         pub fn overwrite_key(
             origin: OriginFor<T>,
             identifier: VerificationKeyIdentifier,
             key: Vec<u8>,
         ) -> DispatchResult {
-            ensure_root(origin)?;
+            let who = ensure_signed(origin).map_err(|_| Error::<T>::BadOrigin)?;
+            let owner = VerificationKeyOwners::<T>::get(identifier);
+            let previous_key = VerificationKeys::<T>::get(identifier);
+
+            if let Some(owner) = owner {
+                ensure!(who == owner, Error::<T>::NotOwner)
+            };
+
+            if let Some(previous) = previous_key {
+                ensure!(
+                    previous.len() == key.len(),
```

Comment:
This check is only due to the guessed value of `MaximumVerificationKeyLength` or there is some other reason?

Reply:
yes they very well could be - for different proof system for example.

I though about that (returning the difference in deposits) it seemed like adding additional computational overhead for very little gain, we can revisit this though. 

When would a key be overwritten:
- relation changed (unlikey)
-  proof system changed (very unlikey in a production system)
- ... ?

- Thoughts @pmikolajczyk41 ?

Reply:
Implemented reimbursement / top up when overwriting. So you can overwrite with a key of any length